main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala

```diff
@@ -255,12 +255,41 @@ object TypeReduction {
    */
   private def isSubtype(tpe1: Type, tpe2: Type)(implicit flix: Flix): Boolean = {
     (tpe2, tpe1) match {
-      case (_, Type.Null) => true // Null is a sub-type of every other type
       case (t1, t2) if t1 == t2 => true
       case (Type.Cst(TypeConstructor.Native(clazz1), _), Type.Cst(TypeConstructor.Native(clazz2), _)) => clazz1.isAssignableFrom(clazz2)
       case (Type.Cst(TypeConstructor.Native(clazz), _), Type.Cst(TypeConstructor.Str, _)) => clazz.isAssignableFrom(classOf[String])
       case (Type.Cst(TypeConstructor.Native(clazz), _), Type.Cst(TypeConstructor.BigInt, _)) => clazz.isAssignableFrom(classOf[BigInteger])
       case (Type.Cst(TypeConstructor.Native(clazz), _), Type.Cst(TypeConstructor.BigDecimal, _)) => clazz.isAssignableFrom(classOf[java.math.BigDecimal])
+      // Null is a sub-type of every Java object and non-primitive Flix type
+      case (Type.Cst(TypeConstructor.Native(_), _), Type.Null) => true
+      case (tpe, Type.Null) if !isPrimitive(tpe) => true
+    }
+  }
+
+  /**
+   * Returns true iff the given type tpe is a Flix primitive.
+   */
+  private def isPrimitive(tpe: Type)(implicit flix: Flix): Boolean = {
+    tpe match {
+      case baseType: Type.BaseType => true // to check
+      case Type.Cst(TypeConstructor.Unit, _) => true
+      case Type.Cst(TypeConstructor.Null, _) => true
+      case Type.Cst(TypeConstructor.Bool, _) => true
+      case Type.Cst(TypeConstructor.Char, _) => true
+      case Type.Cst(TypeConstructor.Float32, _) => true
+      case Type.Cst(TypeConstructor.Float64, _) => true
+      case Type.Cst(TypeConstructor.BigDecimal, _) => true
```

Comment:
In this situation, we're actually concerned about Java primitives, not Flix primitives, which are different. So we should only be looking for the types that correspond to Java primitives (https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html)

Reply:
I see. You mean in this function we are only concerned that correspond to Flix primitives, right?
For the above it will default to false.

Reply:
Don't we have Float32 ~ Float (Java), Float64 ~ Double..?
I don't exactly get the point of the four highlighted lines otherwise. Char, to me, corresponds to a Java primitive and is a Flix primitive. I probably had in mind that Java primitives would be included in Flix primitives and that function would return true for all of these, I did not include Effect or Unit though.

Reply:
I only meant to highlight BigInteger :)

Reply:
Ok I see, thank you.
I considered BigInt as long also.